### PR TITLE
Hhh 11147

### PR DESCRIPTION
**Pull request by sebersole**
https://hibernate.atlassian.net/browse/HHH-11147

Adds the ability to use enhanced entities as a "proxy", meaning being able to return an entity knowing just its identifier using a new interceptor implementation.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+ */
+package org.hibernate.bytecode.enhance.spi.interceptor;
+
+import org.hibernate.engine.spi.PersistentAttributeInterceptor;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+
+/**
+ * @author Steve Ebersole
+ */
+public abstract class AbstractInterceptor implements PersistentAttributeInterceptor, SessionAssociableInterceptor {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: just a minor, `PersistentAttributeInterceptor` can be removed, `SessionAssociableInterceptor` already extends `PersistentAttributeInterceptor`

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/BytecodeLazyAttributeInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Set;
+
+import org.hibernate.Incubating;
+import org.hibernate.engine.spi.PersistentAttributeInterceptor;
+
+/**
+ * @author Steve Ebersole
+ */
+@Incubating
+public interface BytecodeLazyAttributeInterceptor extends PersistentAttributeInterceptor, SessionAssociableInterceptor {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/BytecodeLazyAttributeInterceptor.java]: as in the previous comment, `PersistentAttributeInterceptor` can be removed.

File: hibernate-core/src/test/java/org/hibernate/test/bytecode/enhancement/lazy/group/ComplexFetchGroupTest.java [hunk 3]
```diff
@@ -25,80 +25,200 @@
...
+@TestForIssue(jiraKey = "HHH-11223")
+@RunWith(BytecodeEnhancerRunner.class)
+@CustomEnhancementContext(EnhancerTestContext.class)
+//@FailureExpected( jiraKey = "HHH-11223" )
 public class ComplexFetchGroupTest extends BaseNonConfigCoreFunctionalTestCase {
 
 	@Test
-	public void testNonOwningOneToOneAccess() {
+	public void testLoadNonOwningOneToOne() {
+		// Test loading D and accessing E
```
Comment (dreab8) [inline hibernate-core/src/test/java/org/hibernate/test/bytecode/enhancement/lazy/group/ComplexFetchGroupTest.java]: I think we can remove this comment

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java [hunk 4]
```diff
@@ -191,36 +208,54 @@ protected void entityIsTransient(MergeEvent event, Map copyCache) {
 		LOG.trace( "Merging transient instance" );
 
 		final Object entity = event.getEntity();
-		final EventSource source = event.getSession();
+		if ( entity instanceof PersistentAttributeInterceptable ) {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java]: this check can be removed

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java [hunk 5]
```diff
@@ -253,17 +253,27 @@ private Object proxyOrLoad(
...
+					return managed;
+				}
 			}
 
 			// we won't create proxies, but if there is already one associated with
 			// the PC we should use it
-			final Object proxy = persistenceContext.getProxy( keyToLoad );
-			if ( proxy != null ) {
-				return proxy;
+			if ( persister.hasProxy() ) {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java]: inside this block  `persister.getEntityMetamodel().getBytecodeEnhancementMetadata().isEnhancedForLazyLoading()` have to be `true` so `persister.hasProxy()` so if i do no miss something based on the implementation 
`public boolean hasProxy() {
		// skip proxy instantiation if entity is bytecode enhanced
		return entityMetamodel.isLazy() && !entityMetamodel.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading();
	}`
it will be always `false`

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java [hunk 6]
```diff
@@ -1171,6 +1181,7 @@ public final Object internalLoad(
 				event.setEntityId( null );
 				event.setInstanceToLoad( null );
 				event.setResult( null );
+				event.setShouldUnwrapProxy( unwrapProxy );
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: I think the set should be moved before the `fireLoad( event, type );` at line 1174.

**Comment by gbadner** [comment]
Rebased master with commit that skips failing OffsetTime tests, and force-pushed.

**Comment by sebersole** [comment]
The merge tests seemed wrong to me.  I changed the tests and all 3 pass - I was kind of surprised the "merge detached uninitialized bytecode proxy" passed as-is.

Anyway, fixed tests, squashed and force pushed.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java [hunk 7]
```diff
@@ -4279,6 +4307,50 @@ public Object load(Serializable id, Object optionalObject, LockOptions lockOptio
...
+					entity,
+					session,
+					LockOptions.READ
+			);
+
+			final LazyAttributeLoadingInterceptor interceptor = enhancementMetadata.injectInterceptor(
+					entity,
+					proxyInterceptor.getEntityKey().getIdentifier(),
+					session
+			);
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java]: IIUC ```PojoEntityTuplizer#afterInitialize``` has already injected the ```LazyAttributeLoadingInterceptor```. I don't think it hurts anything, but seems redundant.

File: hibernate-core/src/main/java/org/hibernate/Hibernate.java [hunk 8]
```diff
@@ -187,7 +202,10 @@ public static boolean isPropertyInitialized(Object proxy, String propertyName) {
 
 		if ( entity instanceof PersistentAttributeInterceptable ) {
 			PersistentAttributeInterceptor interceptor = ( (PersistentAttributeInterceptable) entity ).$$_hibernate_getInterceptor();
-			if ( interceptor != null && interceptor instanceof LazyAttributeLoadingInterceptor ) {
+			if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
+				return false;
+			}
+			if ( interceptor instanceof LazyAttributeLoadingInterceptor ) {
 				return ( (LazyAttributeLoadingInterceptor) interceptor ).isAttributeLoaded( propertyName );
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/Hibernate.java]: Currently, ```Hibernate.isPropertyInitialized( activity, "instruction" )``` will return ```true``` if the attribute is initialized to an uninitialized proxy.

At the same time, ```Hibernate.isInitialized( activity.getInstruction() )``` will return ```false```.

I find this confusing. IMO, the return values for both should be ```false```.

File: hibernate-core/src/main/java/org/hibernate/boot/spi/SessionFactoryOptions.java [hunk 9]
```diff
@@ -308,4 +308,8 @@ default boolean areJPACallbacksEnabled() {
 		return true;
 	}
 
+	/**
+	 * Can bytecode-enhanced entity classes be used as a "proxy"?
+	 */
+	boolean isEnhancementAsProxyEnabled();
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/boot/spi/SessionFactoryOptions.java]: Since this is a new SPI method, shouldn't this be added as a default method returning ```false```?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java (hunk 1 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: Also, ```AbstractSessionAssociableInterceptor``` might be a better name.

Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: Should this class implement ```Serializable```?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		return session;
+	}
+
+	@Override
+	public void setSession(SharedSessionContractImplementor session) {
+		this.session = session;
+		if ( session != null && !allowLoadOutsideTransaction ) {
+			this.allowLoadOutsideTransaction = session.getFactory().getSessionFactoryOptions().isInitializeLazyStateOutsideTransactionsEnabled();
+			if ( this.allowLoadOutsideTransaction ) {
+				this.sessionFactoryUuid = session.getFactory().getUuid();
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: Is it possible that this ```AbstractInterceptor``` will become associated with a ```Session``` from a different ```SessionFactory```? If so, ```sessionFactoryUuid``` won't be updated if ```allowLoadOutsideTransaction``` == ```true```.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	@Override
+	public SharedSessionContractImplementor getLinkedSession() {
+		return session;
+	}
+
+	@Override
+	public void setSession(SharedSessionContractImplementor session) {
+		this.session = session;
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: Should further checks be done when  ```this.session != null && session != null```, similar to what is done in [```AbstractPersistentCollection#setCurrentSession```](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/collection/internal/AbstractPersistentCollection.java#L678-L698)?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+			SharedSessionContractImplementor session) {
+		super( entityName, session );
+
+		this.identifierAttributeNames = identifierAttributeNames;
+		assert identifierAttributeNames != null;
+
+		this.nonAggregatedCidMapper = nonAggregatedCidMapper;
+		assert nonAggregatedCidMapper != null || identifierAttributeNames.size() == 1;
+
+		this.entityKey = entityKey;
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java]: ```entityKey``` must be non-null?

File: hibernate-core/src/main/java/org/hibernate/Hibernate.java (hunk 8 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/Hibernate.java]: I see that in ```FetchGraphTest#testLoadNonOwningOneToOne```, that this is intended.

It would be helpful to add documentation to this method that a return value of ```true``` could indicate that the property can be initialized to an uninitialized collection or proxy.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java [hunk 13]
```diff
@@ -4279,6 +4307,50 @@ public Object load(Serializable id, Object optionalObject, LockOptions lockOptio
...
+			}
+
+			return interceptor.readObject(
+					entity,
+					nameOfAttributeBeingAccessed,
+					value
+			);
+		}
+
+		throw new IllegalStateException(  );
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java]: Probably won't happen, but there should be a description of the exception.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/EntityPersister.java [hunk 14]
```diff
@@ -798,7 +814,11 @@ Object createProxy(Serializable id, SharedSessionContractImplementor session)
 	EntityTuplizer getEntityTuplizer();
 
 	BytecodeEnhancementMetadata getInstrumentationMetadata();
-	
+
+	default BytecodeEnhancementMetadata getBytecodeEnhancementMetadata() {
+		return getInstrumentationMetadata();
+	}
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/EntityPersister.java]: ```getBytecodeEnhancementMetadata()``` does not seem to be used anywhere. Should it be removed?

File: hibernate-core/src/main/java/org/hibernate/Hibernate.java (hunk 8 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/Hibernate.java]: I disagree.  The contract of isPropertyInitialized is whether that value is available on the class

File: hibernate-core/src/main/java/org/hibernate/boot/spi/SessionFactoryOptions.java (hunk 9 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/boot/spi/SessionFactoryOptions.java]: Maybe

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java (hunk 10 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: Have you run into this before?  This is exactly the code as it happened before this work.  I see what you are saying, but I think that has nothing to do with this work.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java (hunk 12 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java]: Ofc

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java (hunk 5 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java]: * `#hasProxy` - can it produce HibernateProxy instances
* `#isEnhancedForLazyLoading` - is the entity enhanced

File: hibernate-core/src/main/java/org/hibernate/Hibernate.java (hunk 8 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/Hibernate.java]: Unless I am off base, the same should be true in 4.x if an object is enhanced and has a proxy for an association.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java (hunk 11 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java]: It would not hurt I guess.  But again this was taken directly from the existing interceptor code.  Have you seen trouble with that?  I think this is another one of "not specific to this work"

File: hibernate-core/src/main/java/org/hibernate/persister/entity/EntityPersister.java (hunk 14 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/EntityPersister.java]: No.  Actually the other way around - that is the method we should use.  I got caught not finishing what I intended there.  I have changed that (and some other things) and will push soon

**Comment by gbadner** [comment]
Supersedes https://github.com/hibernate/hibernate-orm/pull/2864

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java [hunk 15]
```diff
@@ -93,62 +160,94 @@ public Object forceInitialize(Object target, String attributeName) {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java]: can we remove the` BytecodeLogger.LOGGER.tracef`? I think the `forceInitialize` method that is called after does it

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java [hunk 16]
```diff
@@ -309,7 +309,7 @@ private Object proxyOrLoad(
 			persistenceContext.addEntity(
 					entity,
 					Status.MANAGED,
-					ArrayHelper.filledArray( LazyPropertyInitializer.UNFETCHED_PROPERTY, Object.class, persister.getPropertyTypes().length ),
+					new Object[ persister.getPropertyTypes().length ],
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java]: Why is this change necessary? Is it possible that a loaded property will be mistaken for null?
